header/p2p/server.go

```diff
@@ -151,43 +159,74 @@ func (serv *ExchangeServer) requestHandler(stream network.Stream) {
 // if it exists.
 func (serv *ExchangeServer) handleRequestByHash(hash []byte) ([]*header.ExtendedHeader, error) {
 	log.Debugw("server: handling header request", "hash", tmbytes.HexBytes(hash).String())
+	ctx, span := serverTracer.Start(serv.ctx, "handleRequestByHash", trace.WithAttributes(
+		attribute.String("hash", tmbytes.HexBytes(hash).String()),
+	))
+	defer span.End()
 
-	h, err := serv.store.Get(serv.ctx, hash)
+	h, err := serv.store.Get(ctx, hash)
 	if err != nil {
 		log.Errorw("server: getting header by hash", "hash", tmbytes.HexBytes(hash).String(), "err", err)
+		span.RecordError(err)
 		return nil, err
 	}
+
+	span.AddEvent("fetched_header_from_store", trace.WithAttributes(
+		attribute.String("hash", tmbytes.HexBytes(hash).String()),
+		attribute.Int64("height", h.Height)),
+	)
+	span.SetStatus(codes.Ok, "")
 	return []*header.ExtendedHeader{h}, nil
 }
 
 // handleRequest fetches the ExtendedHeader at the given origin and
 // writes it to the stream.
 func (serv *ExchangeServer) handleRequest(from, to uint64) ([]*header.ExtendedHeader, error) {
+	ctx, cancel := context.WithTimeout(serv.ctx, time.Second*5)
+	defer cancel()
+
+	ctx, span := serverTracer.Start(ctx, "handleRequest", trace.WithAttributes(
+		attribute.Int64("from", int64(from)),
+		attribute.Int64("to", int64(to))))
+	defer span.End()
+
 	if from == uint64(0) {
+		span.AddEvent("handling_head_request")
 		log.Debug("server: handling head request")
 		head, err := serv.store.Head(serv.ctx)
 		if err != nil {
 			log.Errorw("server: getting head", "err", err)
+			span.RecordError(err)
 			return nil, err
 		}
+
+		span.AddEvent("fetched_head", trace.WithAttributes(
+			attribute.String("hash", head.Hash().String()),
+			attribute.Int64("height", head.Height)),
+		)
+		span.SetStatus(codes.Ok, "")
 		return []*header.ExtendedHeader{head}, nil
 	}
 
 	if to-from > serv.Params.MaxRequestSize {
 		log.Errorw("server: skip request for too many headers.", "amount", to-from)
+		span.RecordError(header.ErrHeadersLimitExceeded)
 		return nil, header.ErrHeadersLimitExceeded
 	}
 	log.Debugw("server: handling headers request", "from", from, "to", to)
-	ctx, cancel := context.WithTimeout(serv.ctx, time.Second*5)
-	defer cancel()
 	headersByRange, err := serv.store.GetRangeByHeight(ctx, from, to)
 	if err != nil {
+		span.RecordError(err)
 		if errors.Is(err, context.DeadlineExceeded) {
 			log.Warnw("server: requested headers not found", "from", from, "to", to)
 			return nil, header.ErrNotFound
 		}
 		log.Errorw("server: getting headers", "from", from, "to", to, "err", err)
 		return nil, err
 	}
+
+	span.AddEvent("fetched_range_of_headers", trace.WithAttributes(
```

Comment:
can we also add field for start:end ?

Reply:
Added on span creation. Should I move it here?